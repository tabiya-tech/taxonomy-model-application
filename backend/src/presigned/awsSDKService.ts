import {createPresignedPost, PresignedPost, PresignedPostOptions} from "@aws-sdk/s3-presigned-post";
import {S3Client} from "@aws-sdk/client-s3";

export function s3_getPresignedPost(region: string, bucketName: string, folderName: string, maxFileSize: number, expires: number): Promise<PresignedPost> {
  const client = new S3Client({region: region});
  const params: PresignedPostOptions = {
    Bucket: bucketName,
    // The ${filename} will be automatically replaced by the name of the file provided, see
    // https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-HTTPPOSTForms.html#sigv4-HTTPPOSTFormFields
    Key: folderName + "/${filename}",
    Fields: {
      acl: 'private',
    },
    Conditions: [
      {bucket: bucketName},
      {acl: "private"},
      ["starts-with", "$key", folderName + "/"],
      ["content-length-range", 0, maxFileSize],
    ],
    Expires: expires,
  };
  return createPresignedPost(client, params);
}