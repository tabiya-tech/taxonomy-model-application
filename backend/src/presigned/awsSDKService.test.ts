// ####
// Set up the mock for the S3Client
jest.mock("@aws-sdk/client-s3", () => {
  return {
    S3Client: jest.fn()
  }
});
// Set up the mock for the createPresignedPost
jest.mock("@aws-sdk/s3-presigned-post", () => {
  return {
    createPresignedPost: jest.fn()
  }
});
// ####

import {S3Client} from "@aws-sdk/client-s3";
import {createPresignedPost, PresignedPost} from "@aws-sdk/s3-presigned-post";
import {s3_getPresignedPost} from "./awsSDKService";

describe('test getPresignedPost()', () => {

  it("should call the S3Client with the correct parameters", async () => {
    // GIVEN a region, bucketName, folder, maxFileSize, and expires
    const givenRegion = "foo";
    const givenBucketName = "bar";
    const givenFolder = "baz";
    const givenMaxFileSize = 100;
    const givenExpires = 1000;
    // AND the S3.createPreSignedPost will successfully return a presigned post object
    const givenMockPost: PresignedPost = {
      url: "foo/url",
      fields: {"key1": "value", "key2": "value2"}
    };
    (createPresignedPost as jest.Mock).mockResolvedValueOnce(givenMockPost);

    // WHEN the getPresignedPost() is called with the given parameters
    const actualPostData: PresignedPost = await s3_getPresignedPost(givenRegion, givenBucketName, givenFolder, givenMaxFileSize, givenExpires);

    // THEN it should use the correct region
    expect(S3Client).toHaveBeenCalledWith({region: givenRegion});
    // AND the given region, bucketName, folder, maxFileSize, and expires should be passed to the createPresignedPost()
    const expectedArg = (createPresignedPost as jest.Mock).mock.calls[0][1]
    expect(expectedArg).toMatchSnapshot();
    // AND it should return the presigned post data that was returned from the createPresignedPost()
    expect(actualPostData).toEqual(givenMockPost);
  });
});