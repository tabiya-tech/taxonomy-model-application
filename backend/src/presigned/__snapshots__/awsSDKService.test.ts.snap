// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test getPresignedPost() should call the S3Client with the correct parameters 1`] = `
{
  "Bucket": "bar",
  "Conditions": [
    {
      "bucket": "bar",
    },
    {
      "acl": "private",
    },
    [
      "starts-with",
      "$key",
      "baz/",
    ],
    [
      "content-length-range",
      0,
      100,
    ],
  ],
  "Expires": 1000,
  "Fields": {
    "acl": "private",
  },
  "Key": "baz/\${filename}",
}
`;
